mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp

```diff
@@ -317,23 +317,51 @@ class DotConverter final : public OpConversionPattern<migraphx::DotOp> {
 
     // check batch dimension. Tosa matmul only allow a single dimension for it,
     // add reshape ops to flatten and restore the original dimension.
-    ArrayRef<int64_t> orgOutDims = outputTy.getShape();
+    SmallVector<int64_t, 5> orgOutDims(outputTy.getShape());
     RankedTensorType newOutType = RankedTensorType::get(orgOutDims, elementTy);
     size_t outRank = orgOutDims.size();
-
-    if (outRank != 3) { // A, B, Out have the same rank. rank=2 assumes batch=1
-      ArrayRef<int64_t> orgDimsA = in_A.getType().cast<ShapedType>().getShape();
-      ArrayRef<int64_t> orgDimsB = in_B.getType().cast<ShapedType>().getShape();
-      int64_t batchSize = 1;
+    SmallVector<int64_t, 5> orgDimsA(
+        in_A.getType().cast<ShapedType>().getShape());
+    SmallVector<int64_t, 5> orgDimsB(
+        in_B.getType().cast<ShapedType>().getShape());
+    size_t rankA = orgDimsA.size();
+    size_t rankB = orgDimsB.size();
+
+    // A, B, Out have the same rank. rank=2 assumes batch=1.
+    // Here handling special cases.
+    if (outRank != 3 || rankA != rankB ||
```

Comment:
Is this meant to work for outRank > 4 ?

Reply:
Yes, it works for that cases as well. i.e., it first flatten the batch dimension and potentially handle the broadcast when outRank > 4